### PR TITLE
[numpy][Do Not Review]add op insert

### PR DESCRIPTION
**Pull request by JiangZhaoh**
## Description ##
add op: insert
ONLY forward

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/numpy/multiarray.py [hunk 1]
```diff
@@ -7030,6 +7030,87 @@ def einsum(*operands, **kwargs):
...
+    array([1.  , 1.  , 7.13, 0.  , 2.  , 2.  , 3.  , 3.  ])
+
+    >>> x = np.arange(8).reshape(2, 4)
+    >>> idx = np.array([1, 3], dtype=np.int32)
+    >>> np.insert(x, idx, np.array([999]), axis=1)
+    array([[  0., 999.,   1.,   2., 999.,   3.],
+           [  4., 999.,   5.,   6., 999.,   7.]])
+    """
+    return _mx_nd_np.insert(arr, obj, values, axis=axis)
+
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: One more blank line below.

File: src/operator/numpy/np_insert_op-inl.h [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+        }
+      }
+      origin_idx[i] = cnt;
+    }
+  }
+};
+
+template<typename xpu>
+void NumpyInsertCompute(const nnvm::NodeAttrs& attrs,
+                          const OpContext& ctx,
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: alignment, same for all other places.

File: python/mxnet/symbol/numpy/_symbol.py [hunk 3]
```diff
@@ -2487,6 +2487,102 @@ def ceil(x, out=None, **kwargs):
...
 
 
+@set_module('mxnet.symbol.numpy')
+def insert(arr, obj, values, axis=None):
+    """
+    Insert values along the given axis before the given indices.
+
+    Parameters
+    ----------
+    arr : ndarray
```
Comment (hgt312) [inline python/mxnet/symbol/numpy/_symbol.py]: Use `_Symbol` instead of `ndarray`, remove examples.

File: src/operator/numpy/np_insert_op-inl.h [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+    }
+
+    int N = arr.shape_[axis];
+    mxnet::TShape newshape(arr.shape_);
+    size_t indices_len = 0;
+    int start = 0, stop = 0, step = 0;
+
+    // get and check indices from slice or sequence of ints
+    if (inputs.size() == 3U) {
+        indices_len = inputs[insert_::kObj].shape_.Size();
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: 2-space indentations. Check for all other places as well.

File: src/operator/numpy/np_insert_op-inl.h [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+};
+
+struct IndicesModify {
+  template<typename IType>
+  MSHADOW_XINLINE static void Map(int i, IType* indices, const int* order) {
+    indices[order[i]] += i;
+  }
+};
+
+struct AssignInsertZero {
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: just use existing `set_zero` kernel

File: src/operator/numpy/np_insert_op-inl.h [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+                        const OpContext& ctx,
+                        const std::vector<TBlob>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+
+  const NumpyInsertParam& param = nnvm::get<NumpyInsertParam>(attrs.parsed);
+  CHECK_EQ(inputs.size(),
+          (param.step.has_value() || param.int_ind.has_value()) ? 2U : 3U);
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: alignment:
```c++
  CHECK_EQ(inputs.size(),
           (param.step.has_value() || param.int_ind.has_value()) ? 2U : 3U);
```

File: src/operator/numpy/np_insert_op-inl.h [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+    indices_len = static_cast<size_t>(seq_cnt);
+  }
+
+  int numnew = 0;  // output.shape[axis] - arr.shape[axis]
+  int index = 0;  // modified index
+  mxnet::TShape val_newshape(arr.shape_.ndim(), -1);
+  // modify values's ndim to arr's ndim, for broadcast easily later
+  // e.g. value shape: (2,) arr shape: (3, 2) => value shape: (1, 2)
+  for (int i = values.shape_.ndim() - 1, j = arr.shape_.ndim() - 1;
+       i >= 0 || j >= 0; --i, --j) {
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]:
```c++
  for (int i = values.shape_.ndim() - 1, j = arr.shape_.ndim() - 1;
       i >= 0 || j >= 0;
       --i, --j) {
```

File: src/operator/numpy/np_insert_op-inl.h [hunk 8]
```diff
@@ -0,0 +1,632 @@
...
+  }
+  mshadow::Shape<10> k_valshape = GetKernelShape<10>(values.shape_);
+  for (int i = 0 ; i < values.shape_.ndim() ; ++i) {
+    k_valshape[i] = values.shape_[i];
+  }
+  MXNET_ASSIGN_REQ_SWITCH(req[insert_::kOut], req_type, {
+    MSHADOW_TYPE_SWITCH(outputs[insert_::kOut].type_flag_, DType, {
+      MSHADOW_TYPE_SWITCH((inputs.size() == 3U) ?
+                           inputs[insert_::kObj].type_flag_ :
+                           mshadow::DataType<int64_t>::kFlag, IType, {
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]:
```c++
      MSHADOW_TYPE_SWITCH((inputs.size() == 3U) ?
                          inputs[insert_::kObj].type_flag_ :
                          mshadow::DataType<int64_t>::kFlag, IType, {
```

File: src/operator/numpy/np_insert_op-inl.h [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+            CHECK((values.shape_[i] == 1) || (values.shape_[i] == sz));
+          }
+          size_t temp_storage_bytes, temp_mem_size;
+          temp_storage_bytes = SortByKeyWorkspaceSize<IType, int, xpu>(indices_len, false, true);
+          temp_mem_size = indices_len * sizeof(IType) * 2 +
+                          indices_len * sizeof(int) +
+                          outshape[axis] * sizeof(int) * 2 +
+                          temp_storage_bytes;
+          Tensor<xpu, 1, char> temp_mem =
+                  ctx.requested[0].get_space_typed<xpu, 1, char>(Shape1(temp_mem_size), s);
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: 2-space indentations.

File: src/operator/numpy/np_insert_op-inl.h [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+          IType* indices_ptr = reinterpret_cast<IType*>(temp_mem.dptr_);
+          IType* sorted_indices_ptr = reinterpret_cast<IType*>
+                                      (temp_mem.dptr_ + indices_len * sizeof(IType));
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]:
```c++
          IType* sorted_indices_ptr =
            reinterpret_cast<IType*>(temp_mem.dptr_ + indices_len * sizeof(IType));
```

File: src/operator/numpy/np_insert_op.cc [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                           std::vector<int> *in_type,
+                           std::vector<int> *out_type) {
+  const NumpyInsertParam& param = nnvm::get<NumpyInsertParam>(attrs.parsed);
+  int insize = (param.step.has_value() || param.int_ind.has_value()) ? 2 : 3;
+  CHECK_EQ(in_type->size(), insize);
+  CHECK_EQ(out_type->size(), 1U);
+  if (insize == 3) {
+    CHECK_NE((*in_type)[2], -1) << "Index type must be set for insert operator\n";
+    CHECK(((*in_type)[2] == mshadow::DataType<int64_t>::kFlag)
+          || ((*in_type)[2] == mshadow::DataType<int32_t>::kFlag))
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op.cc]:
```c++
    CHECK(((*in_type)[2] == mshadow::DataType<int64_t>::kFlag) ||
          ((*in_type)[2] == mshadow::DataType<int32_t>::kFlag))
```

File: src/operator/numpy/np_insert_op.cc [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+      << "Index type only support int32 or int64.\n";
+  }
+  TYPE_ASSIGN_CHECK(*out_type, 0, (*in_type)[0]);
+  TYPE_ASSIGN_CHECK(*out_type, 0, (*in_type)[1]);
+  TYPE_ASSIGN_CHECK(*in_type, 0, (*out_type)[0]);
+  return (*in_type)[0] != -1;
+}
+
+bool NumpyInsertShape(const nnvm::NodeAttrs& attrs,
+                            mxnet::ShapeVector *in_shape,
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op.cc]: alignment:
```c++
bool NumpyInsertShape(const nnvm::NodeAttrs& attrs,
                      mxnet::ShapeVector *in_shape,
                      mxnet::ShapeVector *out_shape) {
```

File: src/operator/numpy/np_insert_op.cc [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    } else if (i >= 0) {
+      CHECK_EQ(valshape[i], 1) << "index exceed limits.";
+    } else {
+      val_newshape[j] = 1;
+    }
+  }
+  valshape.assign(val_newshape.begin(), val_newshape.end());
+
+  if (param.int_ind.has_value() ||
+    (in_shape->size() == 3U && objShape.ndim() == 0)) {
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op.cc]: alignment

Comment (haojin2) [reply src/operator/numpy/np_insert_op.cc]:
```c++
  if (param.int_ind.has_value() ||
      (in_shape->size() == 3U && objShape.ndim() == 0)) {
```

File: src/operator/numpy/np_insert_op-inl.h [hunk 14]
```diff
@@ -0,0 +1,638 @@
...
+  }
+};
+
+struct AssignId {
+  MSHADOW_XINLINE static void Map(int i, int* order) {
+    order[i] = i;
+  }
+};
+
+struct IndicesModify {
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: use `range_fwd` kernel from `src/operator/tensor/init_op.h`  instead

File: src/operator/numpy/np_insert_op-inl.h [hunk 15]
```diff
@@ -0,0 +1,638 @@
...
+  MSHADOW_XINLINE static void Map(int i, int64_t* indices,
+                                  int N, const int64_t* obj) {
+    indices[i] = obj[i];
+    if (indices[i] < 0) {
+      indices[i] += static_cast<int64_t>(N);
+    }
+  }
+};
+
+struct AssignId {
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: use `range_fwd` kernel from `src/operator/tensor/init_op.h`  instead

File: src/operator/numpy/np_insert_op_tensor.cc [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+#include "./np_insert_op-inl.h"
+#include "./np_insert_op_tensor-inl.h"
+
+namespace mxnet {
+namespace op {
+
+DMLC_REGISTER_PARAMETER(NumpyInsertParam);
+
+bool NumpyInsertTensorType(const nnvm::NodeAttrs& attrs,
+                                  std::vector<int> *in_type,
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op_tensor.cc]: alignment.

File: src/operator/numpy/np_insert_op-inl.h [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+        k_outshape, k_valshape, index, numnew,
+        val_strides, old_val_strides, arr_strides, out_strides,
+        axis, moveaxis, req);
+    });
+  });
+}
+
+template<typename xpu, int ndim>
+void InsertSizeOneTensorImpl(mshadow::Stream<xpu> *s, const TBlob& output,
+                            const TBlob& arr, const TBlob& values,
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: alignment

File: src/operator/numpy/np_insert_op-inl.h [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+        k_outshape, k_valshape, N, index.dptr<int64_t>(), numnew,
+        val_strides, old_val_strides, arr_strides, out_strides,
+        axis, moveaxis, req);
+    });
+  });
+}
+
+template<typename xpu, int ndim>
+void InsertSequenceImpl(mshadow::Stream<xpu> *s, const TBlob& output,
+                       const TBlob& arr, const TBlob& values,
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op-inl.h]: alignment

File: src/operator/numpy/np_insert_op_slice-inl.h [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+      // If insert use single index and 'value' is inputed as numerical parameter
+      values = TBlob(ctx.requested[0].get_space_typed<xpu, 1, VType>(Shape1(1), s));
+      Fill(s, values, kWriteTo, param.val.value());
+    });
+  }
+
+  if (indices_len == 1) {
+    MXNET_NDIM_SWITCH(outshape.ndim(), ndim, {
+      InsertScalerObj<xpu, ndim>(s, outputs[out_pos], arr, values,
+                                  mxnet_op::calc_stride(arr.shape_.get<ndim>()),
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op_slice-inl.h]: alignment

File: src/operator/numpy/np_insert_op_slice-inl.h [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+    Kernel<IndicesModify, xpu>::Launch(s, indices_len, indices_ptr, order_ptr);
+
+    mxnet_op::Kernel<mxnet_op::set_zero, xpu>::Launch(s, outshape[axis], is_insert);
+    Kernel<SetIsInsert, xpu>::Launch(s, indices_len, indices_ptr, is_insert);
+
+    Kernel<SetOriginValuesIdx, xpu>::Launch(s, indices_len, indices_ptr, origin_idx);
+    Kernel<SetOriginArrIdx, xpu>::Launch(s, outshape[axis], is_insert, origin_idx);
+    MXNET_NDIM_SWITCH(outshape.ndim(), ndim, {
+      InsertSequenceImpl<xpu, ndim>(s, outputs[out_pos], arr, values,
+                                   mxnet_op::calc_stride(arr.shape_.get<ndim>()),
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op_slice-inl.h]: alignment

File: src/operator/numpy/np_insert_op_tensor-inl.h [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+      // If insert use single index and 'value' is inputed as numerical parameter
+      values = TBlob(ctx.requested[0].get_space_typed<xpu, 1, VType>(Shape1(1), s));
+      Fill(s, values, kWriteTo, param.val.value());
+    });
+  }
+  if (inputs[obj_pos].shape_.ndim() == 0) {
+    // 'obj' is tensor and the tensor's ndim is 0, also need to moveaxis
+    MXNET_NDIM_SWITCH(outshape.ndim(), ndim, {
+      InsertSizeOneTensorImpl<xpu, ndim>(s, outputs[out_pos], arr, values,
+                                        mxnet_op::calc_stride(arr.shape_.get<ndim>()),
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op_tensor-inl.h]: alignment

File: src/operator/numpy/np_insert_op_tensor-inl.h [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+                                        mxnet_op::calc_stride(old_valshape.get<ndim>()),
+                                        mxnet_op::calc_stride(outshape.get<ndim>()),
+                                        outshape.get<ndim>(), values.shape_.get<ndim>(),
+                                        dtype, vtype, req[out_pos], axis, inputs[obj_pos],
+                                        numnew, N, outshape.Size(), true);
+    });
+  } else if (indices_len == 1) {
+    MXNET_NDIM_SWITCH(outshape.ndim(), ndim, {
+      InsertSizeOneTensorImpl<xpu, ndim>(s, outputs[out_pos], arr, values,
+                                        mxnet_op::calc_stride(arr.shape_.get<ndim>()),
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op_tensor-inl.h]: alignment

File: src/operator/numpy/np_insert_op_tensor-inl.h [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+    Kernel<IndicesModify, xpu>::Launch(s, indices_len, indices_ptr, order_ptr);
+
+    mxnet_op::Kernel<mxnet_op::set_zero, xpu>::Launch(s, outshape[axis], is_insert);
+    Kernel<SetIsInsert, xpu>::Launch(s, indices_len, indices_ptr, is_insert);
+
+    Kernel<SetOriginValuesIdx, xpu>::Launch(s, indices_len, indices_ptr, origin_idx);
+    Kernel<SetOriginArrIdx, xpu>::Launch(s, outshape[axis], is_insert, origin_idx);
+    MXNET_NDIM_SWITCH(outshape.ndim(), ndim, {
+      InsertSequenceImpl<xpu, ndim>(s, outputs[out_pos], arr, values,
+                                   mxnet_op::calc_stride(arr.shape_.get<ndim>()),
```
Comment (haojin2) [inline src/operator/numpy/np_insert_op_tensor-inl.h]: alignment

**Pull request merged**